pandas/core/frame.py

```diff
@@ -3862,10 +3863,27 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
         -------
         dataframe : DataFrame
         """
-        inplace = validate_bool_kwarg(inplace, 'inplace')
         if not isinstance(keys, list):
             keys = [keys]
 
+        # collect elements from keys that are scalar (or tuples for MultiIndex)
+        col_labels = [x for x in keys if is_scalar(x) or isinstance(x, tuple)]
+        if any(x not in self for x in col_labels):
+            # if there are any labels that are invalid, we raise a KeyError
+            missing = [x for x in col_labels if x not in self]
+            raise KeyError('{}'.format(missing))
+
+        # raise error for elements of keys that are neither valid column keys
+        # or an array of allowed type
+        if any(not (is_scalar(x) or isinstance(x, tuple)) and
+               not isinstance(x, (ABCSeries, ABCIndexClass, ABCMultiIndex,
+                                  list, np.ndarray)) for x in keys):
+            raise TypeError('keys may only contain a combination of the '
+                            'following: valid column keys, Series, Index, '
+                            'MultiIndex, list or np.ndarray')
```

Comment:
This looks like a lot of iterations over the `keys` (iterating over `col_labels` is equivalent time-complexity-wise).  Could we have a single for-loop where we iterate over `keys` and fail fast?

Reply:
Not bad, but doesn't handle the duplicate column keys yet.

Reply:
I don't quite see the difference between what you wrote, and what I did.  Where do you handle duplicate column keys in your changes above?

Reply:
Nevermind. Forgot that this case had been removed now by this PR

Reply:
In that case, unless there are other objections, I would suggest using the single for-loop instead.